beamer/check/commands.py

```diff
@@ -0,0 +1,190 @@
+)
+@click.password_option(
+    "--password",
+    type=str,
+    default="",
+    prompt=False,
+    help="The password needed to unlock the keystore file.",
+)
+@click.option(
+    "--rpc-file",
+    type=click.Path(exists=True, file_okay=True, dir_okay=False, path_type=Path),
+    required=True,
+    help="Path to the RPC config file.",
+)
+@click.option(
+    "--artifacts-dir",
+    type=click.Path(exists=True, file_okay=False, dir_okay=True, path_type=Path),
+    required=True,
+    help="The directory containing deployment artifacts.",
+)
+@click.option(
+    "--abi-dir",
+    type=click.Path(exists=True, file_okay=False, dir_okay=True, path_type=Path),
+    required=True,
+    help="Path to the directory with contract ABIs.",
+)
+@click.option(
+    "--output",
+    type=click.Path(file_okay=True, dir_okay=False, path_type=Path),
+    required=True,
+    help="Path to store the invalidation info at, which can be later used for verification.",
+)
+@click.option(
+    "--count",
+    type=click.IntRange(1),
+    default=1,
+    show_default=True,
+    help="Number of invalidations to create, per (PROOF_SOURCE, PROOF_TARGET) pair.",
+)
+@click.argument("proof-source", type=ChainIdParam())
+@click.argument("proof-target", type=ChainIdParam(), nargs=-1, required=True)
+def initiate_l1_invalidations(
+    keystore_file: Path,
+    password: str,
+    rpc_file: Path,
+    artifacts_dir: Path,
+    abi_dir: Path,
+    output: Path,
+    count: int,
+    proof_source: ChainId,
+    proof_target: tuple[ChainId],
+) -> None:  # pylint: disable=unused-argument
+    """Initiate L1 invalidations from chain PROOF_SOURCE to each PROOF_TARGET.
+    PROOF_SOURCE and PROOF_TARGET are chain IDs."""
+    beamer.util.setup_logging(log_level="DEBUG", log_json=False)
+
+    account = beamer.util.account_from_keyfile(keystore_file, password)
+    log.info("Loaded keystore file", address=account.address)
+
+    abi_manager = ABIManager(abi_dir)
+    deployment = beamer.artifacts.load(artifacts_dir, proof_source)
+
+    rpc_info = beamer.util.load_rpc_info(rpc_file)
+    url = rpc_info[proof_source]
+    w3 = make_web3(url, account)
+    assert w3.eth.chain_id == proof_source
+    log.info("Connected to RPC", url=url)
+
+    fill_manager = obtain_contract(w3, abi_manager, deployment, "FillManager")
+    request_manager = obtain_contract(w3, abi_manager, deployment, "RequestManager")
+    finality_period = request_manager.functions.chains(proof_source).call()[0]
+
+    invalidations = []
+    for chain_id in proof_target:
+        for n in range(count):
+            invalidation = _make_invalidation(
+                fill_manager, proof_source, chain_id, finality_period
+            )
+            invalidations.append(invalidation)
+
+            # Write the output immediately so that we have at least partial data
+            # in case an error occurs during invalidation for the next chain.
+            data = apischema.serialize(invalidations)
```

Comment:
Related to my comment above. I think we can do a bit better here. 

Imagine we have 5 chains and 3 invalidations each. This means we have 60 invalidations in total (If I'm not mistaken). Ideally, at the end we have a json file with 60 invalidation information, right? 

Imagine somewhere in the middle at invalidation 32, we exit due to TransactionFailed (we can only speculate, but my gut feeling tells me that with 60 transactions over 5 chains it's not unlikely that this happens once in a while). 

Now we have a partial invalidation file, which we cannot extend by resuming the command. It can get a bit tricky now. Of course, we could run a modified command with the chain combinations which haven't been executed yet, which requires some manual work. Then we can only write it into a new file and merge the two by hand. 
Lastly where the command broke we end up with a partial state, where we might have to again modify the command cause what we are gonna essentially do is running over two nested for loops where the outer chain at which the command broke will be partially done and partially not.

It's a good thinking to save the state at each run, but it would be good if the command is able to "resume" the work in case it crashed. Otherwise I see unnecessary manual work coming towards us.

One idea could be, similarly as with the state file, to read the output if the file exists and skip all invalidations which are already there and extend the file.

Reply:
OTOH, if we want to make new invalidations we need to give a new output file (or remove the old one).

Reply:
Heh, when thinking about this, I was like "well, cat is a good enough tool". :rofl: 
(In the sense, just like you mentioned above, you could stitch together multiple output file to have everything in one, but the downside is, of course, you would need to manually track where you left off when runnig additional commands.)

> Imagine we have 5 chains and 5 invalidations each. This means we have 20 invalidations in total (If I'm not mistaken). >Ideally, at the end we have a json file with 20 invalidation information, right?

The answer here depends on how you count.
If you count also the proof source chain into the 5 chains then you will get 20, yes. (4 proof target chains times 5 invalidations each).
If the 5 chains are all proof target chains, then you will get 25 invalidations in total. (5 proof target chains times 5 invalidations each).

> In any case, if our goal is to have a file containing all invalidations from one chain to all others, then we should have  the tools to create this file in case of a crash

Yeah, I can see the benefit of just being able to resume where you left off in case something failed. The data is already there. I will extend it so that if the output file exists already, we continue so that one does not need to adjust the command. :+1:

Reply:
Made it so the command reads the output file if it exists and continues.